components/encryption/src/log_file.rs

```diff
@@ -0,0 +1,326 @@
+/// ```
+#[derive(Debug)]
+pub struct LogFile {
+    base: PathBuf,
+    name: String,
+    // Avoid reopening the file every time when append. And lock the file to ensure that
+    // the file will not be read and written at the same time.
+    file_lock: Mutex<Option<File>>,
+}
+
+impl LogFile {
+    pub fn new(base: PathBuf, name: &str) -> Result<LogFile> {
+        let log_file = LogFile {
+            base,
+            name: name.to_owned(),
+            file_lock: Mutex::new(None),
+        };
+
+        let file_dict = FileDictionary::default();
+        log_file.write(file_dict)?;
+        Ok(log_file)
+    }
+
+    pub fn open(base: PathBuf, name: &str) -> Result<LogFile> {
+        let log_file = LogFile {
+            base,
+            name: name.to_owned(),
+            file_lock: Mutex::new(None),
+        };
+
+        let file_dict = log_file.recovery()?;
+        log_file.write(file_dict)?;
+        Ok(log_file)
+    }
+
+    /// Rewrite the log file to reduce file size and reduce the time of next recovery.
+    pub fn write(&self, file_dict: FileDictionary) -> Result<()> {
+        let mut append_file = self.file_lock.lock().unwrap();
+        let origin_path = self.base.join(&self.name);
+        let mut tmp_path = origin_path.clone();
+        tmp_path.set_extension(format!("{}.{}", thread_rng().next_u64(), TMP_FILE_SUFFIX));
+        let mut tmp_file = OpenOptions::new()
+            .create(true)
+            .write(true)
+            .open(&tmp_path)
+            .unwrap();
+
+        let file_dict_bytes = file_dict.write_to_bytes()?;
+        let header = Header::new(&file_dict_bytes);
+        tmp_file.write_all(&header.to_bytes())?;
+        tmp_file.write_all(&file_dict_bytes)?;
+        tmp_file.sync_all()?;
+
+        // Replace old file with the tmp file aomticlly.
+        rename(&tmp_path, &origin_path)?;
+        let base_dir = File::open(&self.base)?;
+        base_dir.sync_all()?;
+        let file = std::fs::OpenOptions::new().append(true).open(origin_path)?;
+        append_file.replace(file);
+
+        Ok(())
+    }
+
+    /// Recovery from the log file and return `FileDictionary`.
+    pub fn recovery(&self) -> Result<FileDictionary> {
+        let mut _lock = self.file_lock.lock().unwrap();
+        let mut f = OpenOptions::new()
+            .read(true)
+            .open(self.base.join(&self.name))?;
+        let mut file_dict = FileDictionary::default();
+        let mut buf = Vec::new();
+        f.read_to_end(&mut buf)?;
+        let mut remained = buf.as_slice();
+
+        // parse the file dictionary
+        let (_, content) = Header::parse(remained)?;
+        remained.consume(Header::SIZE + content.len());
+        file_dict.merge_from_bytes(content)?;
+
+        // parse the remained records
+        while !remained.is_empty() {
+            let (used_size, file_name, mode) = parse_next_record(remained)?;
```

Comment:
If we encounter a corrupted record, we can probably check how far away it is from the end of file. If it is likely to be the tail record (within 1k of the file end?) we can ignore the corrupted record. Otherwise return error.

Reply:
After discussion, the current logic is: Return after encountering an error, and manually intervene in processing